File: src/main/java/me/miki/shindo/management/mods/impl/JumpCircleMod.java
package me.miki.shindo.management.mods.impl;

import me.miki.shindo.Shindo;
import me.miki.shindo.injection.interfaces.IMixinMinecraft;
import me.miki.shindo.injection.interfaces.IMixinRenderManager;
import me.miki.shindo.management.color.AccentColor;
import me.miki.shindo.management.event.EventTarget;
import me.miki.shindo.management.event.impl.EventJump;
import me.miki.shindo.management.event.impl.EventRender3D;
import me.miki.shindo.management.event.impl.EventUpdate;
import me.miki.shindo.management.language.TranslateText;
import me.miki.shindo.management.mods.Mod;
import me.miki.shindo.management.mods.ModCategory;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.util.Vec3;
import org.lwjgl.opengl.GL11;

import java.util.ArrayList;
import java.util.List;

public class JumpCircleMod extends Mod {

    private final List<JumpCircle> circles = new ArrayList<JumpCircle>();
    private boolean jumping;

    public JumpCircleMod() {
        super(TranslateText.JUMP_CIRCLE, TranslateText.JUMP_CIRCLE_DESCRIPTION, ModCategory.RENDER);
    }

    private static double createAnimation(double value) {
        return Math.sqrt(1.0 - Math.pow(value - 1.0, 2.0));
    }

    @EventTarget
    public void onUpdate(EventUpdate event) {

        if (jumping && mc.thePlayer.onGround) {
            jumping = false;
            circles.add(new JumpCircle(mc.thePlayer.getPositionVector()));
        }

        circles.removeIf(JumpCircle::update);
    }

    @EventTarget
    public void onJump(EventJump event) {
        jumping = true;
    }

    @EventTarget
    public void onRender3D(EventRender3D event) {

        AccentColor currentColor = Shindo.getInstance().getColorManager().getCurrentColor();

        GL11.glPushMatrix();
        GL11.glEnable(3042);
        GL11.glDisable(3008);
        GL11.glDisable(2884);
        GL11.glDisable(3553);
        GL11.glShadeModel(7425);

        for (JumpCircle circle : circles) {
            GL11.glBegin(8);
            for (int i = 0; i <= 360; i += 5) {

                float red = (float) (currentColor.getInterpolateColor().getRGB() >> 16 & 255) / 255.0F;
                float green = (float) (currentColor.getInterpolateColor().getRGB() >> 8 & 255) / 255.0F;
                float blue = (float) (currentColor.getInterpolateColor().getRGB() & 255) / 255.0F;

                Vec3 pos = circle.pos();
                double x = Math.cos(Math.toRadians(i)) * createAnimation(1.0 - circle.getAnimation(((IMixinMinecraft) mc).getTimer().renderPartialTicks)) * 0.7;
                double z = Math.sin(Math.toRadians(i)) * createAnimation(1.0 - circle.getAnimation(((IMixinMinecraft) mc).getTimer().renderPartialTicks)) * 0.7;
                GL11.glColor4d(red, green, blue, 0.6 * circle.getAnimation(((IMixinMinecraft) mc).getTimer().renderPartialTicks));
                GL11.glVertex3d(pos.xCoord + x, pos.yCoord + (double) 0.2f, pos.zCoord + z);
                GL11.glColor4d(red, green, blue, 0.2 * circle.getAnimation(((IMixinMinecraft) mc).getTimer().renderPartialTicks));
                GL11.glVertex3d(pos.xCoord + x * 1.4, pos.yCoord + (double) 0.2f, pos.zCoord + z * 1.4);
            }
            GL11.glEnd();
        }

        GL11.glEnable(3553);
        GL11.glDisable(3042);
        GL11.glEnable(3008);
        GL11.glShadeModel(7424);
        GL11.glEnable(2884);
        GL11.glPopMatrix();
        GlStateManager.resetColor();
    }

    private class JumpCircle {

        private final Minecraft mc = Minecraft.getMinecraft();

        private final Vec3 vector;
        private int tick;
        private int prevTick;

        public JumpCircle(Vec3 vector) {
            this.vector = vector;
            this.prevTick = 20;
            this.prevTick = this.tick = 20;
        }

        public double getAnimation(float pt) {
            return ((float) this.prevTick + (float) (this.tick - this.prevTick) * pt) / 20.0f;
        }

        public boolean update() {
            this.prevTick = this.tick;
            return this.tick-- <= 0;
        }

        public Vec3 pos() {
            return new Vec3(this.vector.xCoord - ((IMixinRenderManager) mc.getRenderManager()).getRenderPosX(), this.vector.yCoord - ((IMixinRenderManager) mc.getRenderManager()).getRenderPosY(), this.vector.zCoord - ((IMixinRenderManager) mc.getRenderManager()).getRenderPosZ());
        }
    }
}

File: src/main/kotlin/me/miki/shindo/management/mods/impl/JumpCircleMod.kt
package me.miki.shindo.management.mods.impl

import me.miki.shindo.Shindo.Companion.getInstance
import me.miki.shindo.injection.mixin.interfaces.client.IMixinMinecraft
import me.miki.shindo.injection.mixin.interfaces.client.renderer.IMixinRenderManager
import me.miki.client_api.event.EventTarget
import me.miki.shindo.management.event.impl.EventJump
import me.miki.shindo.management.event.impl.EventRender3D
import me.miki.shindo.management.event.impl.EventUpdate
import me.miki.shindo.management.language.TranslateText
import me.miki.shindo.management.mods.Mod
import me.miki.shindo.management.mods.ModCategory
import me.miki.shindo.management.nanovg.font.LegacyIcon
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.util.Vec3
import org.lwjgl.opengl.GL11
import kotlin.math.cos
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt

class JumpCircleMod : Mod(
    TranslateText.JUMP_CIRCLE,
    TranslateText.JUMP_CIRCLE_DESCRIPTION,
    ModCategory.RENDER,
    LegacyIcon.MOD_JUMP_CIRCLE
) {
    private val circles: MutableList<JumpCircle> = ArrayList<JumpCircle>()
    private var jumping = false

    @EventTarget
    fun onUpdate(event: EventUpdate?) {
        if (jumping && mc.thePlayer.onGround) {
            jumping = false
            circles.add(JumpCircle(mc.thePlayer.positionVector))
        }

        circles.removeIf { obj: JumpCircle? -> obj!!.update() }
    }

    @EventTarget
    fun onJump(event: EventJump?) {
        jumping = true
    }

    @EventTarget
    fun onRender3D(event: EventRender3D?) {
        val currentColor = getInstance().colorManager.getCurrentColor()

        GL11.glPushMatrix()
        GL11.glEnable(3042)
        GL11.glDisable(3008)
        GL11.glDisable(2884)
        GL11.glDisable(3553)
        GL11.glShadeModel(7425)

        for (circle in circles) {
            GL11.glBegin(8)
            var i = 0
            while (i <= 360) {
                val red = (currentColor.getInterpolateColor().rgb shr 16 and 255).toFloat() / 255.0f
                val green = (currentColor.getInterpolateColor().rgb shr 8 and 255).toFloat() / 255.0f
                val blue = (currentColor.getInterpolateColor().rgb and 255).toFloat() / 255.0f

                val pos = circle.pos()
                val x: Double =
                    cos(Math.toRadians(i.toDouble())) * createAnimation(1.0 - circle.getAnimation(((mc as IMixinMinecraft).getTimer() as net.minecraft.util.Timer).renderPartialTicks)) * 0.7
                val z: Double =
                    sin(Math.toRadians(i.toDouble())) * createAnimation(1.0 - circle.getAnimation(((mc as IMixinMinecraft).getTimer() as net.minecraft.util.Timer).renderPartialTicks)) * 0.7
                GL11.glColor4d(
                    red.toDouble(),
                    green.toDouble(),
                    blue.toDouble(),
                    0.6 * circle.getAnimation(((mc as IMixinMinecraft).getTimer() as net.minecraft.util.Timer).renderPartialTicks)
                )
                GL11.glVertex3d(pos.xCoord + x, pos.yCoord + 0.2, pos.zCoord + z)
                GL11.glColor4d(
                    red.toDouble(),
                    green.toDouble(),
                    blue.toDouble(),
                    0.2 * circle.getAnimation(((mc as IMixinMinecraft).getTimer() as net.minecraft.util.Timer).renderPartialTicks)
                )
                GL11.glVertex3d(pos.xCoord + x * 1.4, pos.yCoord + 0.2, pos.zCoord + z * 1.4)
                i += 5
            }
            GL11.glEnd()
        }

        GL11.glEnable(3553)
        GL11.glDisable(3042)
        GL11.glEnable(3008)
        GL11.glShadeModel(7424)
        GL11.glEnable(2884)
        GL11.glPopMatrix()
        GlStateManager.resetColor()
    }

    private class JumpCircle(private val vector: Vec3) {
        private val mc: Minecraft = Minecraft.getMinecraft()

        private var tick = 20
        private var prevTick = 20

        fun getAnimation(pt: Float): Double {
            return ((this.prevTick.toFloat() + (this.tick - this.prevTick).toFloat() * pt) / 20.0f).toDouble()
        }

        fun update(): Boolean {
            this.prevTick = this.tick
            return this.tick-- <= 0
        }

        fun pos(): Vec3 {
            return Vec3(
                this.vector.xCoord - (mc.renderManager as IMixinRenderManager).getRenderPosX(),
                this.vector.yCoord - (mc.renderManager as IMixinRenderManager).getRenderPosY(),
                this.vector.zCoord - (mc.renderManager as IMixinRenderManager).getRenderPosZ()
            )
        }
    }

    companion object {
        private fun createAnimation(value: Double): Double {
            return sqrt(1.0 - (value - 1.0).pow(2.0))
        }
    }
}



